### PR TITLE
[JIT] x86/x64 - improved localloc codegen in some cases

### PR DESCRIPTION
**Pull request by TIHan**
**Description**

Should resolve: https://github.com/dotnet/runtime/issues/6546

@BruceForstall and I went through this, and it turns out that it was actually quite simple.

The x86/x64 emitter keeps track of the stack whenever it sees the pattern `sub esp`, but there are cases where we do not want to do this. To prevent it from keeping track, it would move `esp` into a temporary register, do `sub` with that register, and then move it back to `esp`. This causes two extra instructions to be emitted for x86 under this scenario.

The change adds a new instruction, `INS_sub_hide`, that does the same thing as `sub`, but will tell the emitter not to track the stack. We already have other instructions to do this exact thing already with `push` and `pop`:
```cpp
// Does not affect the stack tracking in the emitter
INST5(push_hide,        "push",             IUM_RD, 0x0030FE,     0x000068,     BAD_CODE,     BAD_CODE,     0x000050,    INS_FLAGS_None )
INST5(pop_hide,         "pop",              IUM_WR, 0x00008E,     BAD_CODE,     BAD_CODE,     BAD_CODE,     0x000058,    INS_FLAGS_None )
```

Example diff:
image

**Acceptance Criteria**
- [x] Remove `regTmp` from `genStackPointerConstantAdjustment` and possibly further up.

### CODE REVIEW

**Comment by TIHan** [comment]
@dotnet/jit-contrib This is ready assuming CI succeeds. cc @BruceForstall

**Comment by kunalspathak** [comment]
> @dotnet/jit-contrib This is ready assuming CI succeeds. cc @BruceForstall

Could you paste sample diffs?

**Comment by TIHan** [comment]
Almost all diffs look like this:
image

Was actually surprised, I didn't expect to get x64 diffs but I did find another part where we did the `mov, sub, mov` in a potential x64 path, so I replaced it using `sub_hide`.

**Comment by TIHan** [comment]
@BruceForstall Outerloop did pass, apart from the formatting failures.

File: src/coreclr/jit/codegenxarch.cpp [hunk 1]
```diff
@@ -2646,8 +2632,7 @@ void CodeGen::genLclHeap(GenTree* tree)
         // adds to ESP).
 
         inst_RV(INS_NEG, regCnt, TYP_I_IMPL);
-        regNumber regTmp = tree->GetSingleTempReg();
```
Comment (BruceForstall) [inline src/coreclr/jit/codegenxarch.cpp:2649]: You removed the use of a temporary register here. That means we don't need to request that temporary register from LSRA. You need to figure out how to not request that temp for this specific case in `LinearScan::BuildLclHeap()`.

File: src/coreclr/jit/codegenxarch.cpp [hunk 2]
```diff
@@ -2216,36 +2216,32 @@ void CodeGen::genAllocLclFrame(unsigned frameSize, regNumber initReg, bool* pIni
 //
 // Arguments:
 //    spDelta                 - the value to add to SP. Must be negative or zero.
-//    regTmp                  - x86 only: an available temporary register. If not REG_NA, hide the SP
-//                              adjustment from the emitter, using this register.
+//    canTrack                - x86 only: whether or not to track the SP adjustment
```
Comment (BruceForstall) [inline src/coreclr/jit/codegenxarch.cpp:2219]: `canTrack` seems like an odd name. Makes it sound optional. What not something more like `trackSpAdjustments`?

**Review by BruceForstall — CHANGES_REQUESTED** [review]

File: src/coreclr/jit/codegenxarch.cpp (hunk 2 above)
Comment (TIHan) [reply src/coreclr/jit/codegenxarch.cpp:2219]: Makes sense, though `trackSpAdjustments` is a bit longer, but clearer.

File: src/coreclr/jit/codegenxarch.cpp [hunk 3]
```diff
@@ -2584,6 +2572,7 @@ void CodeGen::genLclHeap(GenTree* tree)
             }
             else
             {
+                assert(tree->AvailableTempRegCount() == 1);
                 regCnt = tree->ExtractTempReg();
```
Comment (BruceForstall) [inline src/coreclr/jit/codegenxarch.cpp:2576]: If there's only one temp reg expected, then use GetSingleTempReg() which already does this assert.
```suggestion
                regCnt = tree->GetSingleTempReg();
```

File: src/coreclr/jit/codegenxarch.cpp [hunk 4]
```diff
@@ -2470,6 +2456,7 @@ void CodeGen::genLclHeap(GenTree* tree)
         }
         else
         {
+            assert(tree->AvailableTempRegCount() == 1);
             regCnt = tree->ExtractTempReg();
```
Comment (BruceForstall) [inline src/coreclr/jit/codegenxarch.cpp:2460]:
```suggestion
            regCnt = tree->GetSingleTempReg();
```

**Review by BruceForstall — COMMENTED** [review]
Couple more comments

File: src/coreclr/jit/codegenxarch.cpp (hunk 3 above)
Comment (TIHan) [reply src/coreclr/jit/codegenxarch.cpp:2576]: Ah, that's good and convenient.

**Comment by TIHan** [comment]
@dotnet/jit-contrib This is ready again.

**Comment by BruceForstall** [comment]
[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=66544&view=ms.vss-build-web.run-extensions-tab)

**Review by BruceForstall — APPROVED** [review]

**Pull request merged**